File: RepoTransBench/source_projects/Python/stjordanis_betterprompt/tests/test_betterprompt_edgecases.py
import betterprompt
import pytest
import math
import os

def test_get_from_dict_or_env_empty_dict_no_env(monkeypatch):
    key = "TEST_MISSING_KEY"
    monkeypatch.delenv(key, raising=False)
    with pytest.raises(ValueError) as excinfo:
        betterprompt.get_from_dict_or_env(key, {})
    assert key in str(excinfo.value)

def test_get_from_dict_or_env_none_dict_env(monkeypatch):
    key = "ENV_ONLY_KEY"
    monkeypatch.setenv(key, "val")
    assert betterprompt.get_from_dict_or_env(key, None) == "val"

def test_get_from_dict_or_env_dict_empty(monkeypatch):
    key = "NO_DICT_KEY"
    monkeypatch.setenv(key, "from_env")
    assert betterprompt.get_from_dict_or_env(key, {}) == "from_env"

def test_openai_class_and_dummy(monkeypatch):
    # Direct test for coverage of openai.DummyOpenAICompletion.create
    result = betterprompt.openai.Completion.create()
    assert isinstance(result, dict)

def test_call_openai_api_key(monkeypatch):
    dummy_logprobs = [0.4, 0.5, 0.6]
    class DummyCompletion:
        @staticmethod
        def create(*args, **kwargs):
            return {"choices": [{"logprobs": {"token_logprobs": dummy_logprobs}}]}
    class DummyOpenAI:
        Completion = DummyCompletion

    # This tests the explicit api_key argument branch
    monkeypatch.setattr(betterprompt, "openai", DummyOpenAI)
    result = betterprompt.call_openai("prompt", api_key="explicit_key")
    assert result == dummy_logprobs

def test_calculate_perplexity_regular():
    token_logprobs = [0, -1, -2]
    perplexity = betterprompt.calculate_perplexity(token_logprobs)
    expected = math.exp(-(sum(token_logprobs) / len(token_logprobs)))
    assert abs(perplexity - expected) < 1e-8

def test_calculate_perplexity_empty():
    # Should return inf if token_logprobs is empty
    result = betterprompt.calculate_perplexity([])
    assert math.isinf(result)

File: RepoTransBench/target_projects/Python/Rust/stjordanis_betterprompt/tests/test_betterprompt_edgecases.rs
use betterprompt::*;
use serial_test::serial;
use std::collections::HashMap;
use std::env;

#[test]
#[serial]
fn test_get_from_dict_or_env_empty_dict_no_env() {
    let key = "TEST_MISSING_KEY";
    // Remove env var (if it exists)
    env::remove_var(key);
    let dict: HashMap<String, String> = HashMap::new();
    let result = get_from_dict_or_env(key, Some(&dict));
    assert!(result.is_err());
    assert!(result.unwrap_err().contains(key));
}

#[test]
#[serial]
fn test_get_from_dict_or_env_none_dict_env() {
    let key = "ENV_ONLY_KEY";
    env::set_var(key, "val");
    let result = get_from_dict_or_env(key, None);
    assert_eq!(result.unwrap(), "val");
    env::remove_var(key);
}

#[test]
#[serial]
fn test_get_from_dict_or_env_dict_empty() {
    let key = "NO_DICT_KEY";
    env::set_var(key, "from_env");
    let dict: HashMap<String, String> = HashMap::new();
    let result = get_from_dict_or_env(key, Some(&dict));
    assert_eq!(result.unwrap(), "from_env");
    env::remove_var(key);
}

#[test]
fn test_openai_class_and_dummy() {
    let val = DummyOpenAICompletion::create();
    assert!(val.contains_key("choices"));
}

#[test]
#[serial]
fn test_call_openai_api_key() {
    // Custom dummy handler
    use serde_json::json;
    let dummy_logprobs = vec![0.4, 0.5, 0.6];
    let custom = move |_args: &[&str], _kwargs: &HashMap<&str, String>| {
        serde_json::from_value(json!({
            "choices": [
                {"logprobs": { "token_logprobs": dummy_logprobs }}
            ]
        })).unwrap()
    };
    set_openai_handler(Box::new(custom));
    let result = call_openai("prompt", None, Some("explicit_key"));
    assert_eq!(result, Some(vec![0.4, 0.5, 0.6]));
    reset_openai_handler();
}

#[test]
fn test_calculate_perplexity_regular() {
    let token_logprobs = vec![0.0, -1.0, -2.0];
    let expected = (-token_logprobs.iter().sum::<f64>() / (token_logprobs.len() as f64)).exp();
    let actual = calculate_perplexity(&token_logprobs);
    assert!((actual - expected).abs() < 1e-8);
}

#[test]
fn test_calculate_perplexity_empty() {
    let result = calculate_perplexity(&[]);
    assert!(result.is_infinite());
}